india_forecast_app/app.py

```diff
@@ -35,18 +35,95 @@ def get_sites(db_session: Session) -> list[SiteSQL]:
     return sites
 
 
-def get_model(asset_type: str):
+def get_generation_data(
+        db_session: Session,
+        sites: list[SiteSQL],
+        timestamp: dt.datetime
+) -> dict[str, pd.DataFrame]:
+    """
+    Gets generation data values for given sites
+
+    Args:
+            db_session: A SQLAlchemy session
+            sites: A list of SiteSQL objects
+            timestamp: The end time from which to retrieve data
+
+    Returns:
+            A Dict containing:
+            - "data": Dataframe containing 15-minutely generation data
+            - "metadata": Dataframe containing information about the site
+    """
+
+    site_uuids = [s.site_uuid for s in sites]
+
+    start = timestamp - dt.timedelta(hours=1)
+    # pad by 1 second to ensure get_pv_generation_by_sites returns correct data
+    end = timestamp + dt.timedelta(seconds=1)
+
+    generation_data = get_pv_generation_by_sites(
+        session=db_session, site_uuids=site_uuids, start_utc=start, end_utc=end
+    )
+
+    # Convert to dataframe
+    generation_df = (pd.DataFrame(
+        [(g.start_utc, g.generation_power_kw, g.site.ml_id) for g in generation_data],
+        columns=["time_utc", "power_kw", "ml_id"]
+    ).pivot(index="time_utc", columns="ml_id", values="power_kw"))
+
+    # Ensure timestamps line up with 3min intervals
+    generation_df.index = generation_df.index.round("3min")
+
+    # Drop any duplicated timestamps
+    generation_df = generation_df[~generation_df.index.duplicated()]
+
+    # xarray (used later) expects columns with string names
+    generation_df.columns = generation_df.columns.astype(str)
+
+    # Handle any missing timestamps
+    contiguous_dt_idx = pd.date_range(start=start, end=end, freq="3min")[:-1]
+    generation_df = generation_df.reindex(contiguous_dt_idx, fill_value=None)
+
+    # Interpolate NaNs
+    generation_df = generation_df.interpolate(method="linear", limit_direction="both")
+
+    # Down-sample from 3 min to 15 min intervals
+    generation_df = generation_df.resample("15min").sum()
```

Comment:
I think this should be `mean`

Reply:
That would mean taking the mean of the 5 3-min readings rather than the total recorded power over the 15-min period - is that right?

Reply:
Yea, the average is slightly better thsnks